### ISSUE TITLE
Support More Time Variance in Cost Parameters

### ISSUE THREAD

**Issue by GabrielKS** [NREL-Sienna/PowerSystems.jl#1150]
Right now, these cost-related parameters can optionally be time series:

- `FuelCurve.fuel_cost`
- `MarketBidCost.incremental_offer_curves`
- `MarketBidCost.decremental_offer_curves`
- `MarketBidCost.initial_input` (about to be added)

This is currently accomplished ("_status quo design_") for each parameter by

1. Making the parameter a `Union{Float64, TimeSeriesKey}` or similar(ish) for `FunctionData`
2. Adding an ad hoc getter and validating setter in [`cost_function_timeseries.jl`](https://github.com/NREL-Sienna/PowerSystems.jl/blob/main/src/models/cost_function_timeseries.jl) to match
3. Teaching the user how to use it (no standard interface for this right now)

It sounds like in the future we will need some more parameters to be time varying, including `CostCurve/FuelCurve.vom_cost`.

I would like to consider a design ("_time series of costs design_") where, instead of cost structs composed of fields that might be `Float64` or `FunctionData` time series, we have time series that can hold whole PSY cost structs, so for instance you could have a time series of `FuelCurve`. Thus, any parameter you want could be time varying automatically.

Reasons to prefer _status quo design_:

1. Different parameters can have different time resolutions, some can be scalar while others are time series, etc.
2. We have fine-grained control over what's allowed to be a time series.
3. It seems to correspond better to how things are done in existing modeling tools.

Reasons to prefer _time series of costs design_:

1. Under _status quo design_, users of `MarketBidCost` have to work directly with `FunctionData`, rather than with the user-friendly `ValueCurve` abstractions. That's pretty bad because aside from this the casual user doesn't even need to know that `FunctionData` exists.
2. Under _status quo design_, lots of code has to be added in `cost_function_timeseries.jl` to support each time varying parameter, and the user interface is kind of clunky, whereas it would be very simple to teach the user to create a time series of `FuelCurve`s if that's what they need.
3. Easy to make anything at all a time series whenever we need that, just implement some HDF serialization methods.

I prototyped _time series of costs design_ in the commits leading up to [here (IS)](https://github.com/NREL-Sienna/InfrastructureSystems.jl/commit/aabe96765b0944077cdf215e2e263f3d2fe8db15) and [here (PSY)](https://github.com/NREL-Sienna/PowerSystems.jl/commit/d75839158a9f30f7e3f503954e238c9b11f4fb1c) a while ago; the deserialization was a little buggy but nothing that couldn't be fixed if we committed to this route.

One alternative ("_better accessors design_") would be to keep time series as individual parameters but come up with a more flexible design to replace what's in [`cost_function_timeseries.jl`](https://github.com/NREL-Sienna/PowerSystems.jl/blob/main/src/models/cost_function_timeseries.jl) so that it's easy to make any parameter a time series without a ton of coding or user interface overhead. Maaaybe the requirement for users to work with `FunctionData` could also be paved over in this layer. I'd have to think about what that would look like.

Thoughts @jd-lara @daniel-thom @pesap?

**Comment by luke-kiernan** [comment]
Revisiting this now that psy6 has landed the explicit-units/cost-curve rework — status check against the two designs discussed here:

### Current Status

  **Achieved:** "cost curves themselves as varying" is real. `CostCurve`/`FuelCurve` now carry their `ValueCurve` type as a parameter (`CostCurve{T,U}`), and `T` can be a time-series-backed curve (`TimeSeriesPiecewiseIncrementalCurve`) or a static curve (`PiecewiseIncrementalCurve`).

  **What we actually built:** Rather than "status quo" (per-field unions) or "time series of costs," we ended up a 3rd option, **paired sibling structs**: `MarketBidCost`/`MarketBidTimeSeriesCost`, `ImportExportCost`/`ImportExportTimeSeriesCost`, `ReserveDemandCurve`/`ReserveDemandTimeSeriesCurve`. No more mixing/matching: either you use the vanilla one where all fields are static, or the time-variant one where all fields are `TimeSeries`-backed objects. This significantly simplifies code flow downstream in PSI.

###   Not fully resolved / worth reopening as follow-ups
  1. **`FuelCurve` didn't get the paired-struct treatment**. `fuel_cost::Union{Float64, TimeSeriesKey}` is still an independent field alongside a separately-parameterized `value_curve::T`.
  2. **Code duplication remains**: there's still no generic "make this time-varying" interface. Each pair (`MarketBid*`, `ImportExport*`, `ReserveDemand*`) hand-rolls its own near-identical resolution/validation logic in that file rather than sharing one mechanism. 
  3. `vom_cost::LinearCurve` — the concrete future need this issue called out ("It sounds like in the future we will need... `CostCurve/FuelCurve.vom_cost`") — is still always static. No time-varying path exists for it.

### Suggested Course of Action
Close this issue and open more targeted ones.

My take on the items in the above "not fully resolved" list: 1. seems worthy of discussion. Is that something we should change? @jd-lara I don't recall why we left `FuelCurve` as-is. 2. would be a deep design change. We have enough of those in psy6 already: defer to later. I 3. could go either way. Is there a project or modeling task where we need time-varying vom curves? If no, then I'd tend towards leaving things as-is.

**Comment by jd-lara** [comment]
> Revisiting this now that psy6 has landed the explicit-units/cost-curve rework — status check against the two designs discussed here:
> 
> ### Current Status
> 
> **Achieved:** "cost curves themselves as varying" is real. `CostCurve`/`FuelCurve` now carry their `ValueCurve` type as a parameter (`CostCurve{T,U}`), and `T` can be a time-series-backed curve (`TimeSeriesPiecewiseIncrementalCurve`) or a static curve (`PiecewiseIncrementalCurve`).
> 
> **What we actually built:** Rather than "status quo" (per-field unions) or "time series of costs," we ended up a 3rd option, **paired sibling structs**: `MarketBidCost`/`MarketBidTimeSeriesCost`, `ImportExportCost`/`ImportExportTimeSeriesCost`, `ReserveDemandCurve`/`ReserveDemandTimeSeriesCurve`. No more mixing/matching: either you use the vanilla one where all fields are static, or the time-variant one where all fields are `TimeSeries`-backed objects. This significantly simplifies code flow downstream in PSI.
> 
> ### Not fully resolved / worth reopening as follow-ups
> 
> 1. **`FuelCurve` didn't get the paired-struct treatment**. `fuel_cost::Union{Float64, TimeSeriesKey}` is still an independent field alongside a separately-parameterized `value_curve::T`.
> 2. **Code duplication remains**: there's still no generic "make this time-varying" interface. Each pair (`MarketBid*`, `ImportExport*`, `ReserveDemand*`) hand-rolls its own near-identical resolution/validation logic in that file rather than sharing one mechanism.
> 3. `vom_cost::LinearCurve` — the concrete future need this issue called out ("It sounds like in the future we will need... `CostCurve/FuelCurve.vom_cost`") — is still always static. No time-varying path exists for it.
> 
> ### Suggested Course of Action
> 
> Close this issue and open more targeted ones.
> 
> My take on the items in the above "not fully resolved" list: 1. seems worthy of discussion. Is that something we should change? [@jd-lara](https://github.com/jd-lara) I don't recall why we left `FuelCurve` as-is. 2. would be a deep design change. We have enough of those in psy6 already: defer to later. I 3. could go either way. Is there a project or modeling task where we need time-varying vom curves? If no, then I'd tend towards leaving things as-is.

Yeah we need to rethink how to store the time varying fuel cost more elegantly
